File: android/app/src/test/java/com/tastebuddy/android/domain/TastePerceptionTest.kt
package com.tastebuddy.android.domain

import kotlinx.serialization.json.*
import org.junit.Assert.*
import org.junit.Test

class TastePerceptionTest {
    @Test fun sharedEvidenceCases() {
        val fixture = AppJson.parseToJsonElement(javaClass.classLoader!!.getResource("TBA/taste-perception-fixtures.json")!!.readText()).jsonObject
        fixture["cases"]!!.jsonArray.forEach { value ->
            val example = value.jsonObject; val label = example.string("id")
            val expected = example["expected"]!!.jsonObject
            val records = example["records"]!!.jsonArray.map { it.jsonObject }.filter {
                it["feedbackCompleted"]!!.jsonPrimitive.boolean && it.string("userId") == "owner"
                    && it.string("confirmationStatus") in listOf("explicit_user_choice", "rule_extracted_statement")
            }.map { row ->
                SensoryObservation(row.string("observationId"), row.string("experienceId"), row.string("foodName"), row.string("observedAt"), "fixture",
                    row.string("kind"), row.string("attribute"), "짠맛", row["value"]!!.jsonPrimitive, row.string("scale"), row.string("target"), row.string("phase"), row.string("phrase"), emptyList(),
                    row["reference"]?.jsonPrimitive?.contentOrNull, mealID = row.string("mealId"), observedAt = row.string("observedAt"), knownAt = row.string("knownAt"),
                    dishKindIDs = row["dishKindIDs"]!!.jsonArray.map { it.jsonPrimitive.content }, restaurantID = row["restaurantID"]?.jsonPrimitive?.contentOrNull, menuItemID = row["menuItemID"]?.jsonPrimitive?.contentOrNull, restaurantName = row["restaurantName"]?.jsonPrimitive?.contentOrNull)
            }
            val cutoff = parseInstant(example["asOf"]?.jsonPrimitive?.contentOrNull)
            val result = TastePerception.build(records, cutoff)
            assertEquals(label, expected["meals"]!!.jsonPrimitive.int, result.evidenceCount)
            assertEquals(label, expected["patterns"]!!.jsonPrimitive.int, result.patterns.size)
            assertEquals(label, expected["changes"]!!.jsonPrimitive.int, result.changes.size)
            assertEquals(label, expected["contrasts"]!!.jsonPrimitive.int, result.contrasts.size)
            assertEquals(label, expected["levels"]!!.jsonArray.map { it.jsonPrimitive.intOrNull }, TasteAxis.entries.map { result.current(it)?.currentLevel })
            expected["recentCounts"]?.let { assertEquals(label, it.jsonArray.map { value -> value.jsonPrimitive.int }, result.current(TasteAxis.salty)?.counts) }
            assertEquals(label, result, TastePerception.build(records + records, cutoff))
            val series = TasteChangeSeries.build(result, emptyList())
            result.changes.forEach { pattern ->
                val row = series.first { it.id == pattern.id }
                assertEquals(listOf(pattern.previousLevel, pattern.currentLevel), row.points.map { it.value })
                assertEquals(listOf(3, 3), row.points.map { it.count })
                assertEquals(pattern.previous.evidenceIDs, row.points.first().evidenceIDs)
                assertEquals(pattern.recent.experienceIDs, row.points.last().experienceIDs)
            }
        }
    }

    @Test fun surveyZeroUnknownAndComparableHistory() {
        val catalog = AppJson.decodeFromString<SurveyCatalog>(javaClass.classLoader!!.getResource("fixtures/taste-survey-golden.json")!!.readText())
        val old = SurveyScoring.result(catalog.items, catalog.items.map { SurveyResponse(it.id, 2) }, "2026-08-01T00:00:00.000Z").snapshot.surveySubmission!!
        val current = SurveyScoring.result(catalog.items, listOf(SurveyResponse(catalog.items[0].id, 0), SurveyResponse(catalog.items[1].id, uncertain = true)), "2026-09-01T00:00:00.000Z").snapshot.surveySubmission!!
        assertTrue(SurveyPerception.points(listOf(current, old.copy(recordedAt = "2026-09-01T00:00:00Z"))).none { it.hasChange })
        val points = SurveyPerception.points(listOf(current, old, current))
        assertEquals(0, points.first { it.axis == catalog.items[0].tasteId }.value)
        assertEquals(1, points.count { it.value != null })
        assertEquals(1, points.count { it.hasChange })
        val row = TasteChangeSeries.build(PerceptionSnapshot(), points).first { it.axis == catalog.items[0].tasteId }
        assertEquals(listOf(2, 0), row.points.map { it.value })
        assertEquals("더 약하게", TasteChangeSeries.changeLabel(row.points))
        assertEquals("비교 부족", TasteChangeSeries.changeLabel(row.points.takeLast(1)))
        assertTrue(row.points.all { it.experienceIDs.isEmpty() })
        val incompatible = old.copy(items = old.items.mapIndexed { index, item -> if (index == 0) item.copy(anchor = item.anchor.copy(conditions = listOf("다른 조리 조건"))) else item })
        assertTrue(SurveyPerception.points(listOf(current, incompatible)).none { it.hasChange })
    }
}

File: scripts/taste-perception.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { buildTastePerception, buildSurveyPerception, buildTasteChangeSeries, tasteChangeLabel } from '../src/lib/tastePerception.mjs';

const fixtures = JSON.parse(readFileSync(new URL('../ios/TasteBuddy/Resources/TBA/taste-perception-fixtures.json', import.meta.url), 'utf8'));
for (const example of fixtures.cases) test(example.id, () => {
  const result = buildTastePerception(example.records, { asOf: example.asOf ?? null, userID: 'owner' });
  assert.equal(result.evidenceCount, example.expected.meals);
  assert.equal(result.patterns.length, example.expected.patterns);
  assert.equal(result.changes.length, example.expected.changes);
  assert.equal(result.contrasts.length, example.expected.contrasts);
  assert.deepEqual(result.axes.map(row => row.current?.currentLevel ?? null), example.expected.levels);
  if (example.expected.recentCounts) assert.deepEqual(result.axes.find(row => row.axis === 'salty').current.counts, example.expected.recentCounts);
  const repeated = buildTastePerception([...example.records, ...example.records], { asOf: example.asOf ?? null, userID: 'owner' });
  assert.deepEqual(repeated, result, '원문 재전송은 근거를 늘리지 않는다');
});

const catalog = JSON.parse(readFileSync(new URL('../ios/TasteBuddy/Resources/Fixtures/taste-survey-golden.json', import.meta.url), 'utf8'));
const original = catalog.cases[0].expected.snapshot.surveySubmission;
test('미각변화 화면은 같은 조건의 두 기간과 원본 근거를 연결하며 회상 0을 보존한다', () => {
  const example = fixtures.cases.find(row => row.expected.changes > 0);
  const model = buildTastePerception(example.records, { userID: 'owner' });
  const pattern = model.changes[0];
  const row = buildTasteChangeSeries(model, { points: [] }).find(row => row.id === pattern.id);
  assert.deepEqual(row.points.map(point => point.value), [pattern.previousLevel, pattern.currentLevel]);
  assert.deepEqual(row.points.map(point => point.count), [3, 3]);
  assert.deepEqual(row.points[0].evidenceIDs, pattern.previous.evidenceIDs);
  assert.deepEqual(row.points[1].experienceIDs, pattern.recent.experienceIDs);
  const old = structuredClone(original); old.recordedAt = '2026-08-01T00:00:00Z';
  const current = structuredClone(original); current.recordedAt = '2026-09-01T00:00:00Z';
  old.responses = [{ itemId: old.items[0].id, selectedValue: 2, uncertain: false }];
  current.responses = [{ itemId: current.items[0].id, selectedValue: 0, uncertain: false }];
  const recall = buildTasteChangeSeries({ patterns: [] }, buildSurveyPerception([old, current]))[0];
  assert.deepEqual(recall.points.map(point => point.value), [2, 0]);
  assert.equal(tasteChangeLabel(recall.points), '더 약하게');
  assert.equal(tasteChangeLabel(recall.points.slice(1)), '비교 부족');
  assert.deepEqual(buildTasteChangeSeries({ patterns: [] }, { points: [] }), []);
});
test('설문 0·모름·미응답과 원래 음식 조건을 보존하고 같은 문항끼리만 비교한다', () => {
  const old = structuredClone(original); old.recordedAt = '2026-08-01T00:00:00.000Z';
  const current = structuredClone(original); current.recordedAt = '2026-09-01T00:00:00.000Z';
  current.responses = [{ itemId: current.items[0].id, selectedValue: 0, uncertain: false },
    { itemId: current.items[1].id, selectedValue: null, uncertain: true, uncertaintyReason: 'cannot_recall' }];
  const before = JSON.stringify([current, old]);
  const result = buildSurveyPerception([current, old, current]);
  assert.equal(result.points.find(row => row.axis === current.items[0].tasteId).value, 0);
  assert.equal(result.points.filter(row => row.value !== null).length, 1);
  assert.equal(result.changes.length, 1);
  assert.equal(JSON.stringify([current, old]), before);
  old.items[0].anchor.conditions[0] = '다른 조리 조건';
  assert.equal(buildSurveyPerception([current, old]).changes.length, 0);
});

test('회상 척도·문항 버전이 다르면 전후 비교를 만들지 않는다', () => {
  const old = structuredClone(original); old.recordedAt = '2026-08-01T00:00:00.000Z';
  const current = structuredClone(original); current.recordedAt = '2026-09-01T00:00:00.000Z';
  current.responses[0].selectedValue = 0;
  old.scale.labels['0'] = '다른 의미';
  assert.equal(buildSurveyPerception([current, old]).changes.length, 0);
  old.scale = current.scale; old.instrument.version = '1.0.0';
  assert.equal(buildSurveyPerception([current, old]).changes.length, 0);
});

 test('같은 시각의 다른 날짜 표기는 전후 변화가 아니다', () => {
  const first = structuredClone(original); first.recordedAt = '2026-09-01T00:00:00.000Z';
  const copy = structuredClone(first); copy.recordedAt = '2026-09-01T00:00:00Z'; copy.responses[0].selectedValue = 0;
  assert.equal(buildSurveyPerception([first, copy]).changes.length, 0);
});
